Add friends likes

ToNyQwW opened the pull request:
Привет) Вроде все сделал по заданию. Но захотел сделать все проверки через аннотации и не знаю, насколько это правильно) Пришлось в методах валидации выбрасывать NotFoundException и, получается, нарушая этим логику методов валидации. Из плюсов -  получились максимально "чистые" классы сервисов и хранилищ. Если так лучше не делать, то переделаю.

Тесты написал только на хранилища ( по минимуму ). Про тесты в задании вообще не было написано, но если надо, то допишу.

ToNyQwW committed:
рефакторинг отступов в коде

ToNyQwW committed:
Добавлен слой storage и перенесена туда логика по работе с сущностями

ToNyQwW committed:
добавлена в сущность user коллекция friends

ToNyQwW committed:
Реализован базовый функционал UserServiceImpl и InMemoryUserStorage

ToNyQwW committed:
добавлена коллекция userIdLikes

ToNyQwW committed:
Реализован InMemoryFilmStorage

ToNyQwW committed:
Реализован FilmServiceImpl

ToNyQwW committed:
доработан UserService

ToNyQwW committed:
реализован UserController

ToNyQwW committed:
реализован FilmController

ToNyQwW committed:
фикс метода addFriend

ToNyQwW committed:
реализован метод getCommonFriends на всех слоях

ToNyQwW committed:
доработан getFilm

ToNyQwW committed:
доработан updateFilm

ToNyQwW committed:
доработан getFilms

ToNyQwW committed:
доработан updateFilm

ToNyQwW committed:
доработаны addLike и removeLike

ToNyQwW committed:
доработаны addLike, removeLike и updateFilm

ToNyQwW committed:
переделан FilmService с помощью валидаций аннотаций

ToNyQwW committed:
добавлен лог в MinReleaseDateValidator

ToNyQwW committed:
рефакторинг аннотаций

ToNyQwW committed:
Изменение типа параметра для Id (сделал Long)

ToNyQwW committed:
доработан UserController

ToNyQwW committed:
доработан UserServiceImpl

ToNyQwW committed:
доработан InMemoryUserStorage

ToNyQwW committed:
доработка в контроллерах

ToNyQwW committed:
добавлены NotFoundExceptions;
перенесены аннотации валидации в интерфейсы;
добавлено выбрасывание NotFoundExceptions

ToNyQwW committed:
работа с выбрасыванием NotFoundException

ToNyQwW committed:
добавил ExceptionHandlerAdvice;
исправлены методы получения популярных фильмов и удаления из друзей

ToNyQwW committed:
добавил handleValidationException для отлавливания исключений связанных с валидацией

ToNyQwW committed:
добавил библиотеку Logbook

ToNyQwW committed:
добавлены тесты на FilmStorage и UserStorage

ToNyQwW committed:
доработки в ExceptionHandlerAdvice

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ExceptionHandlerAdvice.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Map;
+
+@RestControllerAdvice
+public class ExceptionHandlerAdvice {
```
отлично)

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/validation/validFilm/FilmValidator.java:
```diff
@@ -0,0 +1,27 @@
+import ru.yandex.practicum.filmorate.storage.FilmStorage;
+
+@Slf4j
+public class FilmValidator implements ConstraintValidator<ValidFilm, Film> {
```
знаешь в чем минус таких валидаторов - они лишний раз ходят в хранилище. Я согласен, что код в сервисе действительно выглядит очень красиво и чисто, нет никаких проверок на null - но это достигается ценой двойных походов в хранилище (сначала проверить на существование, а потом достать). В случае с мапой это конечно не критично, но позже в работах будут появляться базы данных, и вот в них два раза ходить уже не очень.

И например UserIdValidator - то что id не существует в хранилище не значит же, что он невалидный) Он корректный, просто его нет - а это уже уровень сервисной логики, а не валидации

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -0,0 +1,25 @@
+
+    List<Film> getFilms();
+
+    Film updateFilm(@ValidFilm Film film);
```
в интерфейсах обычно не используются аннотации)
если и используются, то уже в реализациях
интерфейс максимально абстрактен, поэтому не должен накладывать ограничения, кроме типов параметров, списка методов и типов возвращаемых значений

GeorgeGolikov reviewed (commented):
Привет! Очень хорошая работа! Ты молодец, чисто и красиво, как всегда. По поводу валидации оставил комментарии) А тесты в этой работе можно было не писать)

ToNyQwW committed:
Удалил аннотации для валидации (для будущей оптимизации);
Переработан getUser()

ToNyQwW committed:
Избавление от аннотаций в FilmService;

ToNyQwW committed:
переработан updateUser()

ToNyQwW committed:
переработан UserServiceImpl и InMemoryUserStorage

ToNyQwW committed:
переработан FilmServiceImpl и InMemoryFilmStorage

ToNyQwW committed:
фикс тестов

ToNyQwW committed:
переработка логики работы в InMemoryFilmStorage и InMemoryUserStorage

ToNyQwW commented:
Вроде все поправил, теперь нет лишних обращений к коллекции.

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -43,22 +47,37 @@ public List<Film> getFilms() {
+            log.info("Film updated: {}", updatedFilm);
+            return updatedFilm;
+        } catch (NotFoundException e) {
+            log.info(e.getMessage());
```
скорее это уровень error, не info)
и лучше обрабатывать просто Exception - мало ли чего еще могло случиться

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/storage/inmemory/InMemoryFilmStorage.java:
```diff
@@ -31,29 +33,60 @@ public Film addFilm(Film film) {
         return new ArrayList<>(films.values());
     }
 
+    private void containsFilmId(Long filmId) throws NotFoundException {
```
метод звучит как boolean, лучше назвать его throwIfNotExists

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/storage/inmemory/InMemoryFilmStorage.java:
```diff
@@ -31,29 +33,60 @@ public Film addFilm(Film film) {
+    }
+
+    private void checkFilm(Long id, Film film) throws NotFoundException {
+        if (film == null) {
```
такую проверку можно не выносить в отдельный метод)

GeorgeGolikov commented on src/main/java/ru/yandex/practicum/filmorate/storage/inmemory/InMemoryUserStorage.java:
```diff
@@ -32,46 +33,83 @@ public User createUser(User user) {
         return new ArrayList<>(users.values());
     }
 
+    private void containsId(Long userId) throws NotFoundException {
```
по юзеру аналогичные комментарии

GeorgeGolikov reviewed (commented):
Привет! Еще чуть-чуть улучшим, и готово

ToNyQwW committed:
рефакторинг null проверок в Storage классах;
добавление обработки всех исключений в Service классах;
изменение уровня логирования
